### ISSUE TITLE
BatchNorm behaves different in train() and eval()

### ISSUE THREAD

**Issue by yuandong-tian** [pytorch/pytorch#5406]
```python
import torch
import torch.nn as nn
from torch.autograd import Variable

print(torch.__version__)

def get_data(cuda=False):
    if cuda:
        return torch.cuda.FloatTensor(32, 1, 19, 19)
    else:
        return torch.FloatTensor(32, 1, 19, 19)

m = nn.Sequential(
    nn.Conv2d(1, 3, kernel_size=3, padding=1),
    nn.BatchNorm2d(3),
    nn.ReLU()
)

cuda = True
if cuda:
    m.cuda()

m.eval()
print(m._modules['1'].training)

x = Variable(get_data(cuda))
y1 = m.forward(x)

m.train()
print(m._modules['1'].training)
y2 = m.forward(x)

err = (y1 - y2).norm().data[0]
if err > 1e-5:
    print(err)
```

Sample output:
```
0.3.0.post4
False
True
33.47104263305664
```
Also in master:
```
0.4.0a0+0ef1038
False
True
6.883443355560303
```

**Comment by kyunghyuncho** [comment]
https://github.com/zdevito/ATen/blob/master/src/THNN/generic/BatchNormalization.c#L29-L62

these lines may answer the discrepancy?

**Comment by SsnL** [comment]
This is expected. In eval(), BN uses tracked running stats, rather than data stats.

**Comment by soumith** [comment]
this is standard expected behavior. In `eval()` mode, BatchNorm does not rely on batch statistics but uses the `running_mean` and `running_std` estimates that it computed during it's training phase.

This is documented as well:
<img width="690" alt="screen shot 2018-02-25 at 11 02 13 pm" src="https://user-images.githubusercontent.com/1310570/36652947-0ab16f8c-1a80-11e8-9541-935b837b7d27.png">

**Comment by weizequan** [comment]
Hi, if i want to use batch statistic in inference phase, what should i do?

**Comment by soumith** [comment]
@weizequan set your model to `.train()` mode if you want to use batch statistic in inference phase.

**Comment by vaisaxena** [comment]
@soumith you set your model to .eval() mode for inference phase, isn't it ?

**Comment by soumith** [comment]
yes, but @weizequan wants to use the "training" behavoir of using batch statistics rather than running_mean and running_std usage.

**Comment by RizhaoCai** [comment]
> @weizequan set your model to `.train()` mode if you want to use batch statistic in inference phase.

In the testing phase, we cannot set model.train(), right? In practical scenarios, one model accepts one sample as input, therefore batch_size = 1.

**Comment by RizhaoCai** [comment]
> this is standard expected behavior. In `eval()` mode, BatchNorm does not rely on batch statistics but uses the `running_mean` and `running_std` estimates that it computed during it's training phase.
> 
> This is documented as well:
> <img alt="screen shot 2018-02-25 at 11 02 13 pm" width="690" src="https://user-images.githubusercontent.com/1310570/36652947-0ab16f8c-1a80-11e8-9541-935b837b7d27.png">

Hello.  I can understand there is the difference. But, why is the difference so huge. I train a model with TRAINING_SET data with the setting model.train(). Then, I test the model also with the TRAINING_SET data with the setting model.eval and I get very bad performance. It is very weird. 

Could you please tell me why is the difference so huge, and how can I get that better? Like changing batch size, network structures, etc.? I am very confused and do not know how to solve it. What I want is to see reasonable results with the setting model.eval() enabled

**Comment by soumith** [comment]
See my comment here: https://discuss.pytorch.org/t/model-eval-gives-incorrect-loss-for-model-with-batchnorm-layers/7561/3

**Comment by danFromTelAviv** [comment]
does group norm work in the same manner?

**Comment by msachin93** [comment]
Do running means and var get updated in the eval() mode?

**Comment by andreped** [comment]
I have never had the same issues using TensorFlow's batch norm layer, and I observe the same thing as you do in PyTorch. I found that **TensorFlow and PyTorch uses different default parameters for momentum** and epsilon. After changing to TensorFlow's default momentum value from **0.1 -> 0.01**, my model perform just as good in eval model as it does during training. I hope this helps someone, as it was quite annoying for me at the time.

Note that TensorFlow's momentum is 1 - PyTorch's in terms of interpretation.

**Comment by Iman1221** [comment]
Thank you @andreped. I agree with you.
But there is something really weired in pytorch eval mode with batch normalization. 
I set the momentum to 0.01 but it is still not like tensorflow.
Are you setting batch norm to eval mode? Something like:

def set_bn_eval(m):
    if isinstance(m, nn.modules.batchnorm._BatchNorm):
        m.eval()

model.apply(set_bn_eval)

**Comment by andreped** [comment]
> Thank you @andreped. I agree with you.
> But there is something really weired in pytorch eval mode with batch normalization.
> I set the momentum to 0.01 but it is still not like tensorflow.
> Are you setting batch norm to eval mode? Something like:
> 
> def set_bn_eval(m):
> if isinstance(m, nn.modules.batchnorm._BatchNorm):
> m.eval()
> 
> model.apply(set_bn_eval)

Well, the whole point was to be able to deploy the model in test mode. So yes, I modified the momentum, deployed it in test mode, and then it worked. 

When I didn't change the momentum, I was still able to run the model in train mode, and it worked somewhat. However, this is non-ideal for multiple reasons. Simple: model.eval() should be sufficient. 

@Iman1221: But what do you mean like tensorflow? Which model are you attempting to train on which task and data set?

**Comment by gianderiu** [comment]
Hi, is anyone experiencing `NaN` values returned by the layer BatchNorm2d?
My model uses `batch size = 2` because of the size of the GPU memory and it happens in **eval mode** only
The input data is valid, no `NaN` values
```
eps=1e-05
momentum momentum=0.1
```

**Comment by Cchahu** [comment]
> Thank you @andreped. I agree with you. But there is something really weired in pytorch eval mode with batch normalization. I set the momentum to 0.01 but it is still not like tensorflow. Are you setting batch norm to eval mode? Something like:
> 
> def set_bn_eval(m): if isinstance(m, nn.modules.batchnorm._BatchNorm): m.eval()
> 
> model.apply(set_bn_eval)

It seems that your training data is non-stationary. You can try to set track_running_stats=False(Default=True) in your BatchNorm. It worked for me.
